src/basic/virt.c

```diff
@@ -408,9 +408,11 @@ int detect_vm(void) {
         /* x86 xen Dom0 is detected as XEN in hypervisor and maybe others.
          * In order to detect the Dom0 as not virtualization we need to
          * double-check it */
-        if (r == VIRTUALIZATION_XEN && detect_vm_xen_dom0())
-                r = VIRTUALIZATION_NONE;
-        else if (r == VIRTUALIZATION_NONE && other)
+        if (r == VIRTUALIZATION_XEN) {
+                r = detect_vm_xen_dom0();
+                if (r < 0)
+                        return r;
```

Comment:
Hmm, this works, but somehow I'm not too happy about the style. I'd much prefer for this part to look as
```C
if (r == VIRTUALIZATION_XEN) {
           k = detect_xm_xen_dom();
           if (k < 0)
                  return k;
           if (k > 0)
                   r = VIRTUALIZATION_NONE;
} ...
```

The reason is that right now `detect_vm_xen_dom0` assumes that we already detected that we're under xen of some type. But the name of that function does not make that explicit, and anyway it seems that the caller should be making such choices, not the low-level detection function.